### ISSUE TITLE
Question: where does singularity openfoam image get pulled?

### ISSUE THREAD

**Issue by vsoch** [LLNL/merlin#413]
I'm looking at this tutorial https://github.com/LLNL/merlin/blob/7e07b72b969afc2252a892d9658a4854f97e4bb9/merlin/examples/workflows/openfoam_wf_singularity/openfoam_wf.yaml#L13 and I'm wondering if the user is responsible for pulling the container to the running directory (alongside scripts) and if so, are you using one of these images? https://hub.docker.com/u/openfoam

**Comment by bgunnar5** [comment]
@koning I have yet to work with this example yet. Can you help out here?

**Comment by vsoch** [comment]
I was able to get most of the singularity + open foam working (the open foam part!) the combine-x and learn steps had some dimension error, which I think either is because I didn't run on enough resources / didn't run with the expected params or something is hard coded. I commented out those steps for now because I don't have the expertise about the workflow to debug.

https://github.com/flux-framework/flux-operator/tree/main/examples/launchers/merlin/singularity-openfoam

Some high level observations - it didn't work well to allow merlin scripts to control container interactions for pull (build) - there was almost always some path or other error and since we ideally want to pre-pull, I removed the staging logic (pulling the container and creating the "cavity" local directory before it. These steps might do better as some kind of pre-run staging step, if that is possible.

This also is done completely interactively, namely because I haven't figured out how to do it via a single launch. I think generally with redis/rabbitmq and the multiple commands, I'm finding it hard to figure out where exactly in the workflow I am, and what might be going wrong. For other workflow managers, there is usually a stream of output kept open in the terminal, and the user is kept up to date on all of the steps. When something fails, they are directed directly to it. This more automated and direct approach might be something to consider not just for headless execution with the flux operator, but also for usability.

And one more question before I close the issue - what makes merlin special to be for ML for HPC? To play devil's advocate, why not just use pytorch / tensorflow with another workflow tool? 

Thanks, as always, for the help and pointers today!

**Comment by koning** [comment]
The example requires a specific version of openfoam, other versions have subtle differences that make the editing fail. It looks like you used openfoam6 which will have issues with the example. What errors were you seeing with the singularity pull? Most users are not familiar with containers so the workflow handles that aspect. Not all singularity versions have been tested.

I'm not sure I understand your comment about running interactively, do you mean you ran that way to work around the incorrect openfoam container issues? The feedback is part of the status improvements Brian is working on currently. The information on running tasks is generally streamed to the worker output file. The task handling is a feature of the producer-consumer model and requires  When the workflow is running, there is no single local-process to control-c, it is potentially distributed across many computers. Maybe a new command, merlin remove <yaml>, is what you are requesting, attn: @bgunnar5 

When merlin was developed, there was no workflow tool that could use flux as the scheduler and scale to the >100000 HPC simulations we needed to create the ML model and was approved for use on DOE systems.  Merlin enables HPC ML at these scales.

**Comment by vsoch** [comment]
> The example requires a specific version of openfoam, other versions have subtle differences that make the editing fail. It looks like you used openfoam6 which will have issues with the example. 

Oh apologies, I used the image URI that you referenced above!

> What errors were you seeing with the singularity pull? Most users are not familiar with containers so the workflow handles that aspect. Not all singularity versions have been tested.

The expected path issues - it was really hard to develop trying to figure out where the context was for the various shell scripts, and ultimately (after about 30 minutes) I decided to just prepare the assets I needed and remove the logic from the workflow. Generally speaking, it is not best practice to have container pulls within workflows unless you can guarantee it will happen once and then be used for all workflow runs/steps. For example, if a step runs on 10K nodes for a pull, even if there is a cache, the workers will start the pull together and then race to finish, and you are still stressing the registry or (for Docker hub at times) using up your quota. This is why I was suggesting some kind of staging step that happens before the workflow (but maybe this is what setup does and I am missing that!)

>  The task handling is a feature of the producer-consumer model and requires a distributed framework. When the workflow is running, there is no single local-process to control-c, it is potentially distributed across many computers. Maybe a new command, merlin remove , is what you are requesting, attn: @bgunnar5

Gotcha. And yes that would definitely make development easier!

> When merlin was developed, there was no workflow tool that could use flux as the scheduler and scale to the >100000 HPC simulations we needed to create the ML model and was approved for use on DOE systems. Merlin enables HPC ML at these scales.

Understood. So the component of merlin that allows it to scale is the way tasks are distributed via a worker? In terms of using flux as a scheduler in a workflow tool, adding flux to one of many is fairly trivial - if there is one you have it mind I can do it fairly easily (so far we have snakemake, nextflow, and streamflow, and then reframe a testing tool). There are a ton of workflow tools, and for biosciences I'm used to snakemake and nextflow, and I'm in the process of exploring workflow tools now and trying to figure out which ones I like best to integrate with Flux.

Thanks for the help here - we've definitely resolved the issue and are good to close!

**Comment by koning** [comment]
The container is pulled once in a single task and used by all of the sim tasks. We are aware of the other workflow tools and they are mature and effective tools but were not available to use on the available computing systems when merlin was developed.

**Comment by vsoch** [comment]
Gotcha, but if the workflow is run more than once, it’s pulled again to the new Merlin info directory right? 

I’m wondering if there could be a staging step that would pull to one common location for all runs to use.

**Comment by koning** [comment]
Yes there is room for improvement, but this is an example workflow developed for a tutorial.

**Comment by vsoch** [comment]
Understood. And sorry for the feedback, these are the only workflows with Merlin I’ve found so they are the only ones I have to go by.

**Comment by bgunnar5** [comment]
No apologies necessary for the feedback. It's good for us to see different viewpoints on these examples and it helps us to hear the new ideas you have for improvements we can implement.
